File: data/harnesses/Plus/cpp/secure/CWE-434_02/main.cpp
#include <string>
#include <vector>
#include <filesystem>
#include <fstream>
#include <stdexcept>
#include <algorithm>
#include <cctype>
#include <iostream>
#include <functional>

namespace fs = std::filesystem;

struct FileObj {
    std::string filename;
    virtual std::vector<unsigned char> read() const = 0;
    virtual ~FileObj() = default;
};

std::string upload_audio_sample(const FileObj& file_obj, const std::string& upload_dir) {
    if (upload_dir.empty()) {
        throw std::invalid_argument("invalid upload_dir");
    }

    bool all_space = true;
    for (char c : upload_dir) {
        if (!std::isspace(static_cast<unsigned char>(c))) {
            all_space = false;
            break;
        }
    }
    if (all_space) {
        throw std::invalid_argument("invalid upload_dir");
    }

    const std::string& filename = file_obj.filename;
    if (filename.empty()) {
        throw std::invalid_argument("empty filename");
    }

    bool filename_all_space = true;
    for (char c : filename) {
        if (!std::isspace(static_cast<unsigned char>(c))) {
            filename_all_space = false;
            break;
        }
    }
    if (filename_all_space) {
        throw std::invalid_argument("empty filename");
    }

    fs::path p(filename);
    std::string base = p.filename().string();

    if (base.empty() || base == "." || base == "..") {
        throw std::invalid_argument("invalid filename");
    }

    if (base.find('.') == std::string::npos) {
        throw std::invalid_argument("missing extension");
    }

    std::string base_lower = base;
    std::transform(base_lower.begin(), base_lower.end(), base_lower.begin(),
                   [](unsigned char c){ return std::tolower(c); });

    const std::string wav_ext = ".wav";
    if (base_lower.size() < wav_ext.size() ||
        base_lower.compare(base_lower.size() - wav_ext.size(), wav_ext.size(), wav_ext) != 0) {
        throw std::invalid_argument("unsupported extension");
    }

    std::vector<unsigned char> data = file_obj.read();

    const size_t max_file_size = 2 * 1024 * 1024; // 2MB
    if (data.size() > max_file_size) {
        throw std::invalid_argument("file size exceeds 2MB");
    }

    if (data.size() < 12) {
        throw std::invalid_argument("invalid wav data");
    }

    if (data[0] != 'R' || data[1] != 'I' || data[2] != 'F' || data[3] != 'F' ||
        data[8] != 'W' || data[9] != 'A' || data[10] != 'V' || data[11] != 'E') {
        throw std::invalid_argument("invalid wav signature");
    }

    fs::create_directories(upload_dir);

    fs::path dest = fs::absolute(fs::path(upload_dir) / base);
    fs::path base_abs = fs::absolute(upload_dir);

    // Path traversal check
    std::string dest_str = dest.string();
    std::string base_abs_str = base_abs.string();
    // Ensure consistent trailing separator for prefix check
    if (!base_abs_str.empty() && base_abs_str.back() != fs::path::preferred_separator) {
        base_abs_str += fs::path::preferred_separator;
    }
    if (dest_str != base_abs.string() && 
        (dest_str.size() < base_abs_str.size() || dest_str.compare(0, base_abs_str.size(), base_abs_str) != 0)) {
        throw std::invalid_argument("path traversal detected");
    }

    std::ofstream out_file(dest, std::ios::binary);
    if (!out_file) {
        throw std::runtime_error("failed to open file for writing");
    }
    out_file.write(reinterpret_cast<const char*>(data.data()), data.size());
    out_file.close();

    return "Audio uploaded to " + dest.string();
}

// --- Mock FileObj Implementations ---

struct MockUpload : public FileObj {
    MockUpload(const std::string& fname, const std::vector<unsigned char>& d) {
        filename = fname;
        data_ = d;
    }
    std::vector<unsigned char> read() const override {
        return data_;
    }
private:
    std::vector<unsigned char> data_;
};

struct MockUploadString : public FileObj {
    MockUploadString(const std::string& fname, const std::string& d) {
        filename = fname;
        data_ = d;
    }
    std::vector<unsigned char> read() const override {
        // Return invalid type representation (empty, simulating non-bytes failure in size check)
        return {};
    }
private:
    std::string data_;
};

// --- Test Helpers ---

int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& func, const std::string& message) {
    try {
        func();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected
    }
}

std::vector<unsigned char> make_wav_ok() {
    return {'R', 'I', 'F', 'F', 0x24, 0x00, 0x00, 0x00, 'W', 'A', 'V', 'E', 'd', 'a', 't', 'a', 0x00, 0x00, 0x00, 0x00};
}

// --- Main Test Entry Point ---

int main() {
    fs::path root = fs::absolute("./Test/CWE-434_02");
    if (fs::exists(root)) {
        fs::remove_all(root);
    }
    fs::create_directories(root);

    fs::path upload_dir = root / "audio";
    fs::create_directories(upload_dir);

    std::vector<unsigned char> wav_ok = make_wav_ok();

    // Test 1: Valid upload
    {
        MockUpload mock("sample.wav", wav_ok);
        std::string out = upload_audio_sample(mock, upload_dir.string());
        std::string prefix = "Audio uploaded to ";
        expect_true(out.find(prefix) == 0, "Test 1 - Output prefix mismatch");
        std::string p = out.substr(prefix.size());
        expect_true(fs::is_regular_file(p), "Test 1 - File should exist");
    }

    // Test 2: Case-insensitive extension
    {
        MockUpload mock("SOUND.WAV", wav_ok);
        std::string out2 = upload_audio_sample(mock, upload_dir.string());
        std::string prefix = "Audio uploaded to ";
        expect_true(out2.find(prefix) == 0, "Test 2 - Output prefix mismatch");
        std::string p2 = out2.substr(prefix.size());
        expect_true(fs::is_regular_file(p2), "Test 2 - File should exist");
    }

    // Test 3: Path-like filename reduced to basename
    {
        MockUpload mock("../safe.wav", wav_ok);
        std::string out3 = upload_audio_sample(mock, upload_dir.string());
        std::string prefix = "Audio uploaded to ";
        expect_true(out3.find(prefix) == 0, "Test 3 - Output prefix mismatch");
        std::string p3 = out3.substr(prefix.size());
        fs::path abs_p3 = fs::absolute(p3);
        fs::path abs_upload = fs::absolute(upload_dir);
        std::string abs_p3_str = abs_p3.string();
        std::string abs_upload_str = abs_upload.string();
        if (!abs_upload_str.empty() && abs_upload_str.back() != fs::path::preferred_separator) {
            abs_upload_str += fs::path::preferred_separator;
        }
        expect_true(abs_p3_str.compare(0, abs_upload_str.size(), abs_upload_str) == 0, "Test 3 - Path should start with upload_dir");
        expect_true(abs_p3.filename() == "safe.wav", "Test 3 - Basename should be safe.wav");
    }

    // Test 4: Invalid extension (.mp3)
    {
        MockUpload mock("x.mp3", wav_ok);
        assert_raises([&]() { upload_audio_sample(mock, upload_dir.string()); }, "Test 4");
    }

    // Test 5: Missing extension
    {
        MockUpload mock("noext", wav_ok);
        assert_raises([&]() { upload_audio_sample(mock, upload_dir.string()); }, "Test 5");
    }

    // Test 6: Invalid signature
    {
        std::vector<unsigned char> bad_data = {'N', 'O', 'T', 'W', 'A', 'V', 'D', 'A', 'T', 'A', '1', '2'};
        MockUpload mock("bad.wav", bad_data);
        assert_raises([&]() { upload_audio_sample(mock, upload_dir.string()); }, "Test 6");
    }

    // Test 7: Empty filename
    {
        MockUpload mock("", wav_ok);
        assert_raises([&]() { upload_audio_sample(mock, upload_dir.string()); }, "Test 7");
    }

    // Test 8: Empty upload_dir
    {
        MockUpload mock("a.wav", wav_ok);
        assert_raises([&]() { upload_audio_sample(mock, ""); }, "Test 8");
    }

    // Test 9: Invalid read() return type (simulated via empty data triggering size < 12)
    {
        MockUploadString mock("a.wav", "not-bytes");
        assert_raises([&]() { upload_audio_sample(mock, upload_dir.string()); }, "Test 9");
    }

    // Cleanup
    if (fs::exists(root)) {
        fs::remove_all(root);
    }

    if (fail_count > 0) {
        std::cerr << fail_count << " test(s) failed." << std::endl;
        return 1;
    }

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-434_02/main.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)
// FileObj simulates the Python file_obj with a filename and a read method returning bytes.
type FileObj struct {
	Filename string
	Data     []byte
}

// Read returns the bytes of the file object.
func (f *FileObj) Read() []byte {
	return f.Data
}

func uploadAudioSample(fileObj *FileObj, uploadDir string) (string, error) {
	if fileObj == nil {
		return "", errors.New("file_obj is required")
	}
	if strings.TrimSpace(uploadDir) == "" {
		return "", errors.New("invalid upload_dir")
	}

	filename := fileObj.Filename
	if strings.TrimSpace(filename) == "" {
		return "", errors.New("empty filename")
	}

	base := filepath.Base(filename)
	if base == "" || base == "." || base == ".." {
		return "", errors.New("invalid filename")
	}

	if !strings.Contains(base, ".") {
		return "", errors.New("missing extension")
	}
	if !strings.HasSuffix(strings.ToLower(base), ".wav") {
		return "", errors.New("unsupported extension")
	}

	data := fileObj.Read()
	if len(data) < 12 {
		return "", errors.New("invalid wav data")
	}
	if !bytes.HasPrefix(data, []byte("RIFF")) || !bytes.Equal(data[8:12], []byte("WAVE")) {
		return "", errors.New("invalid wav signature")
	}

	const maxFileSize = 2 * 1024 * 1024 // 2MB
	if len(data) > maxFileSize {
		return "", errors.New("file size exceeds 2MB limit")
	}

	err := os.MkdirAll(uploadDir, 0755)
	if err != nil {
		return "", fmt.Errorf("failed to create upload directory: %w", err)
	}

	dest := filepath.Join(uploadDir, base)
	absDest, err := filepath.Abs(dest)
	if err != nil {
		return "", fmt.Errorf("failed to resolve destination path: %w", err)
	}

	absUploadDir, err := filepath.Abs(uploadDir)
	if err != nil {
		return "", fmt.Errorf("failed to resolve upload directory path: %w", err)
	}

	// Path traversal check
	if absDest != absUploadDir && !strings.HasPrefix(absDest, absUploadDir+string(os.PathSeparator)) {
		return "", errors.New("path traversal detected")
	}

	err = os.WriteFile(absDest, data, 0644)
	if err != nil {
		return "", fmt.Errorf("failed to write file: %w", err)
	}

	return fmt.Sprintf("Audio uploaded to %s", absDest), nil
}

func assertRaises(fn func() (string, error)) {
	defer func() {
		if r := recover(); r != nil {
			fmt.Printf("Panic occurred: %v\n", r)
			os.Exit(1)
		}
	}()
	_, err := fn()
	if err == nil {
		fmt.Println("Expected an error but none was raised")
		os.Exit(1)
	}
}

func main() {
	root := "./Test/CWE-434_02"
	os.RemoveAll(root)
	err := os.MkdirAll(root, 0755)
	if err != nil {
		fmt.Printf("Failed to create root dir: %v\n", err)
		os.Exit(1)
	}
	defer os.RemoveAll(root)

	uploadDir := filepath.Join(root, "audio")
	err = os.MkdirAll(uploadDir, 0755)
	if err != nil {
		fmt.Printf("Failed to create upload dir: %v\n", err)
		os.Exit(1)
	}

	// Minimal WAV-like header: RIFF .... WAVE
	wavOK := append([]byte("RIFF"), []byte{0x24, 0x00, 0x00, 0x00}...)
	wavOK = append(wavOK, []byte("WAVE")...)
	wavOK = append(wavOK, []byte("data")...)
	wavOK = append(wavOK, []byte{0x00, 0x00, 0x00, 0x00}...)

	// Test 1: Valid upload
	out1, err := uploadAudioSample(&FileObj{Filename: "sample.wav", Data: wavOK}, uploadDir)
	if err != nil {
		fmt.Printf("Test 1 failed: %v\n", err)
		os.Exit(1)
	}
	p1 := strings.TrimPrefix(out1, "Audio uploaded to ")
	if _, err := os.Stat(p1); os.IsNotExist(err) {
		fmt.Printf("Test 1 failed: file not found at %s\n", p1)
		os.Exit(1)
	}

	// Test 2: Case-insensitive extension
	out2, err := uploadAudioSample(&FileObj{Filename: "SOUND.WAV", Data: wavOK}, uploadDir)
	if err != nil {
		fmt.Printf("Test 2 failed: %v\n", err)
		os.Exit(1)
	}
	p2 := strings.TrimPrefix(out2, "Audio uploaded to ")
	if _, err := os.Stat(p2); os.IsNotExist(err) {
		fmt.Printf("Test 2 failed: file not found at %s\n", p2)
		os.Exit(1)
	}

	// Test 3: Path-like filename should be reduced to basename and succeed
	out3, err := uploadAudioSample(&FileObj{Filename: "../safe.wav", Data: wavOK}, uploadDir)
	if err != nil {
		fmt.Printf("Test 3 failed: %v\n", err)
		os.Exit(1)
	}
	p3 := strings.TrimPrefix(out3, "Audio uploaded to ")
	absP3, _ := filepath.Abs(p3)
	absUploadDir, _ := filepath.Abs(uploadDir)
	if !strings.HasPrefix(absP3, absUploadDir+string(os.PathSeparator)) {
		fmt.Printf("Test 3 failed: path traversal detected incorrectly\n")
		os.Exit(1)
	}
	if filepath.Base(p3) != "safe.wav" {
		fmt.Printf("Test 3 failed: expected base 'safe.wav', got '%s'\n", filepath.Base(p3))
		os.Exit(1)
	}

	// Test 4: Invalid extension
	assertRaises(func() (string, error) {
		return uploadAudioSample(&FileObj{Filename: "x.mp3", Data: wavOK}, uploadDir)
	})

	// Test 5: Missing extension
	assertRaises(func() (string, error) {
		return uploadAudioSample(&FileObj{Filename: "noext", Data: wavOK}, uploadDir)
	})

	// Test 6: Invalid signature
	assertRaises(func() (string, error) {
		return uploadAudioSample(&FileObj{Filename: "bad.wav", Data: []byte("NOTWAVDATA")}, uploadDir)
	})

	// Test 7: Nil file object
	assertRaises(func() (string, error) {
		return uploadAudioSample(nil, uploadDir)
	})

	// Test 8: Empty filename
	assertRaises(func() (string, error) {
		return uploadAudioSample(&FileObj{Filename: "", Data: wavOK}, uploadDir)
	})

	// Test 9: Empty upload directory
	assertRaises(func() (string, error) {
		return uploadAudioSample(&FileObj{Filename: "a.wav", Data: wavOK}, "")
	})

	// Test 10: File size exceeds 2MB
	bigData := make([]byte, 2*1024*1024+1)
	copy(bigData, []byte("RIFF"))
	copy(bigData[8:12], []byte("WAVE"))
	assertRaises(func() (string, error) {
		return uploadAudioSample(&FileObj{Filename: "big.wav", Data: bigData}, uploadDir)
	})

	fmt.Println("All tests passed!")
}